pkg/controllers/cinder/cinder.go

```diff
@@ -0,0 +1,151 @@
+package cinder
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v3/volumetypes"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/openstack-cinder-csi-driver-operator/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/klog/v2"
+)
+
+// This CinderController watches OpenStack and:
+// 1) Installs Cinder CSI driver itself.
+// 2) Creates StorageClass for each volume type provided by Cinder.
+//
+// Note that StorageClasses are not deleted when a volumes type disappears
+// from Cinder.
+
+type CinderController struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Cinder is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Cinder and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CinderController"
+)
+
+func NewCinderController(
```

Comment:
IMO this controller should be more specific rather than a generic Cinder controller that creates StorageClasses **and** dictates what other controllers should run.

I'd recommend we make this controller do only one thing: create StorageClasses. You can take a look at what we did for vSphere [here](https://github.com/openshift/vmware-vsphere-csi-driver-operator/tree/master/pkg/operator/storageclasscontroller).